src/dotnet/MsbuildProject.cs

```diff
@@ -16,44 +19,49 @@ internal class MsbuildProject
     {
         const string ProjectItemElementType = "ProjectReference";
 
-        public ProjectRootElement Project { get; private set; }
+        public ProjectRootElement ProjectRoot { get; private set; }
         public string ProjectDirectory { get; private set; }
 
-        private MsbuildProject(ProjectRootElement project)
+        private ProjectCollection _collection;
+        private List<NuGetFramework> _cachedTfms = null;
+        private Project _cachedEvaluatedProject = null;
+
+        private MsbuildProject(ProjectCollection collection, ProjectRootElement project)
         {
-            Project = project;
-            ProjectDirectory = PathUtility.EnsureTrailingSlash(Project.DirectoryPath);
+            _collection = collection;
+            ProjectRoot = project;
+            ProjectDirectory = PathUtility.EnsureTrailingSlash(ProjectRoot.DirectoryPath);
         }
 
-        public static MsbuildProject FromFileOrDirectory(string fileOrDirectory)
+        public static MsbuildProject FromFileOrDirectory(ProjectCollection collection, string fileOrDirectory)
         {
             if (File.Exists(fileOrDirectory))
             {
-                return FromFile(fileOrDirectory);
+                return FromFile(collection, fileOrDirectory);
             }
             else
             {
-                return FromDirectory(fileOrDirectory);
+                return FromDirectory(collection, fileOrDirectory);
             }
         }
 
-        public static MsbuildProject FromFile(string projectPath)
+        public static MsbuildProject FromFile(ProjectCollection collection, string projectPath)
         {
             if (!File.Exists(projectPath))
             {
                 throw new GracefulException(CommonLocalizableStrings.ProjectDoesNotExist, projectPath);
             }
 
-            var project = TryOpenProject(projectPath);
+            var project = TryOpenProject(collection, projectPath);
             if (project == null)
             {
                 throw new GracefulException(CommonLocalizableStrings.ProjectIsInvalid, projectPath);
             }
 
-            return new MsbuildProject(project);
+            return new MsbuildProject(collection, project);
         }
 
-        public static MsbuildProject FromDirectory(string projectDirectory)
+        public static MsbuildProject FromDirectory(ProjectCollection collection, string projectDirectory)
```

Comment:
I'd love for `collection` to be more descriptive... even `projects` would help as it implies that it's `many of project`.